### ISSUE TITLE
Garbage truck - only picks up on one side

### ISSUE THREAD

**Issue by federicoalvarez-github** [VROOM-Project/vroom#522]
Hello there, I'm working on a thesis regarding garbage trucks and I need to optimize routes.

Is it possible to force the truck to pass by each destination on its right side? Because the lifting is done by a mechanical arm that is mounted facing right side of the truck.

Imagine this case: in a two-way street there are to trashcans, one in front of the other. Currently VROOM considers both visited in one passing, but I need to pass two times, one for each trashcan, and each time the trashcan should be on the right of the truck.

Hope my question is clear enough.

**Comment by jcoupey** [comment]
This is not an optimization problem but rather one at the routing level. The routing engine has to decide whether you can visit two points "in one passing", or need to keep moving, make a turn somewhere and come back to the same street with the correct approach.

If everything is done right at the routing level, then all the information will be embedded in the travel time estimates used by VROOM.

For example with OSRM, this can be achieved using the `approaches` parameters.

**Comment by nilsnolde** [comment]
I'm not entirely sure for Valhalla, I never needed to use that feature, but I think you could model that with `preferred_side` for the input locations (possibly playing with `street_side_tolerance` or `street_side_max_distance`): https://github.com/valhalla/valhalla/blob/master/docs/api/turn-by-turn/api-reference.md#locations

In either case, you'd need to supply a custom `matrix` by requesting the matrix from the routing layer before the call to vroom.

**Comment by federicoalvarez-github** [comment]
Thank you very much Julien for your answer.

Thats exactly what I need. Is it possible to tell VROOM to use that parameter when building the queries to OSRM? If so, where should that go? Thanks again for your prompt reply.

Best regards, Federico.

**Comment by jcoupey** [comment]
There is no way to pass extra parameters out of the box, but if that's what you want as a default you can hard-code the changes directly upon request creation, e.g. [here](https://github.com/VROOM-Project/vroom/blob/7be0b4ffb38015543229bd50e7f7082388c0e62e/src/routing/osrm_routed_wrapper.cpp#L26-L52) for `osrm-routed`.

**Comment by federicoalvarez-github** [comment]
Julien, just a follow-up, I ended up doing a less-elegant but less-complicated solution. Modify the source code, compile and then mantain would have been too much work.

The solution I came to is this:
- Build PHP pages called TRIP and TABLE
- They receive the invocations from VROOM, as if they where an OSRM server
- Internally, they add the APPROACHES variable with the respective CURB values to the QUERY STRING
- Then they consume the real OSRM server, with the modified query string, and return the JSON response
- VROOM receives the new response without knowing it passed through my PHP page

Not the most elegant, but works.

Maybe this helps someone else.

Thanks for your help and putting me in the right track.

Best regards, Federico.
